File: backend/app/schemas/admin_users.py
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.core.rbac import ROLE_NAMES
from app.schemas.admin_ai import validated_quota


UserStatus = Literal["active", "locked", "deleted"]
UserSortBy = Literal[
    "created_at",
    "updated_at",
    "last_login_at",
    "email",
    "full_name",
    "role",
    "status",
]
SortOrder = Literal["asc", "desc"]


def validate_role_value(value: str) -> str:
    role = value.strip()
    if role not in ROLE_NAMES:
        raise ValueError(f"role phải là một trong: {', '.join(sorted(ROLE_NAMES))}.")
    return role


class AdminUserBase(BaseModel):
    full_name: str = Field(..., min_length=1, max_length=160)
    email: EmailStr
    role: str = "user"
    status: UserStatus = "active"
    is_active: bool = True
    email_verified: bool = False
    permissions_override: list[str] = Field(default_factory=list)

    @field_validator("role")
    @classmethod
    def validate_role(cls, value: str) -> str:
        return validate_role_value(value)


class AdminUserCreateRequest(BaseModel):
    full_name: str = Field(..., min_length=1, max_length=160)
    email: EmailStr
    role: str = "user"
    password: Optional[str] = Field(default=None, min_length=6, max_length=128)
    temporary_password: Optional[str] = Field(default=None, min_length=6, max_length=128)
    email_verified: bool = False
    current_quota: Optional[dict[str, Any]] = None

    @field_validator("role")
    @classmethod
    def validate_role(cls, value: str) -> str:
        return validate_role_value(value)


class AdminUserUpdateRequest(BaseModel):
    full_name: Optional[str] = Field(default=None, min_length=1, max_length=160)
    email: Optional[EmailStr] = None
    status: Optional[UserStatus] = None
    is_active: Optional[bool] = None
    email_verified: Optional[bool] = None
    permissions_override: Optional[list[str]] = None


class AdminUserRoleUpdateRequest(BaseModel):
    role: str
    reason: Optional[str] = Field(default=None, max_length=500)

    @field_validator("role")
    @classmethod
    def validate_role(cls, value: str) -> str:
        return validate_role_value(value)


class AdminUserQuotaUpdateRequest(BaseModel):
    current_quota: dict[str, Any] = Field(default_factory=dict)
    reason: Optional[str] = Field(default=None, max_length=500)

    @field_validator("current_quota")
    @classmethod
    def validate_current_quota(cls, value: dict[str, Any]) -> dict[str, int]:
        return validated_quota(value)


class AdminUserReasonRequest(BaseModel):
    reason: Optional[str] = Field(default=None, max_length=500)


class AdminPasswordResetRequest(BaseModel):
    temporary_password: Optional[str] = Field(default=None, min_length=6, max_length=128)
    require_password_change: bool = True


class AdminPasswordResetResponse(BaseModel):
    user_id: str
    temporary_password: str
    password_reset_required: bool
    updated_at: datetime
    audit_event: dict[str, Any]


class AdminUserSummary(BaseModel):
    id: str
    full_name: str
    email: EmailStr
    role: str
    status: UserStatus
    is_active: bool
    email_verified: bool = False
    created_at: datetime
    updated_at: Optional[datetime] = None
    last_login_at: Optional[datetime] = None
    deleted_at: Optional[datetime] = None
    current_quota: Optional[dict[str, Any]] = None


class TokenUsage(BaseModel):
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0


class AdminUserDetail(AdminUserSummary):
    document_count: int = 0
    question_count: int = 0
    conversation_count: int = 0
    ai_request_count: int = 0
    token_usage: TokenUsage = Field(default_factory=TokenUsage)


class AdminUserListResponse(BaseModel):
    items: list[AdminUserSummary]
    total: int
    page: int
    page_size: int
    total_pages: int
    generated_at: datetime


class AdminUserStatisticsResponse(BaseModel):
    total_users: int
    active_users: int
    locked_users: int
    deleted_users: int
    users_created_today: int
    users_created_last_7_days: int
    users_created_last_30_days: int
    active_last_24_hours: int
    active_last_7_days: int
    generated_at: datetime


class AdminUserMutationResponse(BaseModel):
    user: AdminUserDetail
    audit_event: dict[str, Any]

File: frontend/src/types/adminUsers.ts
export type AdminRole = 'super_admin' | 'admin' | 'moderator' | 'support' | 'analyst' | 'user' | 'student' | 'lecturer';
export type AdminUserStatus = 'active' | 'locked' | 'deleted';
export type AdminUserSortBy = 'created_at' | 'updated_at' | 'last_login_at' | 'email' | 'full_name' | 'role' | 'status';
export type SortOrder = 'asc' | 'desc';

export interface TokenUsage {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
}

export interface AdminUserSummary {
  id: string;
  full_name: string;
  email: string;
  role: AdminRole;
  status: AdminUserStatus;
  is_active: boolean;
  email_verified: boolean;
  created_at: string;
  updated_at: string | null;
  last_login_at: string | null;
  deleted_at: string | null;
  current_quota: Record<string, unknown> | null;
  phone_number?: string;
  student_code?: string;
  teacher_code?: string;
  class_name?: string;
  grade?: string;
  subject?: string;
  specialization?: string;
  date_of_birth?: string;
}

export interface UserActivityLog {
  id: string;
  timestamp: string;
  action: string;
  details?: string;
}

export interface AdminUserDetail extends AdminUserSummary {
  document_count: number;
  question_count: number;
  conversation_count: number;
  ai_request_count: number;
  token_usage: TokenUsage;
  enrolled_courses?: string[];
  exercises_done?: number;
  exams_done?: number;
  gpa?: number;
  learning_progress_pct?: number;
  class_count?: number;
  assigned_classes?: string[];
  materials_created?: number;
  questions_created?: number;
  exams_created?: number;
  submissions_graded?: number;
  activity_history?: UserActivityLog[];
}

export interface AdminUserListParams {
  page?: number;
  page_size?: number;
  search?: string;
  role?: AdminRole;
  status?: AdminUserStatus | 'all';
  created_from?: string;
  created_to?: string;
  last_login_from?: string;
  last_login_to?: string;
  sort_by?: AdminUserSortBy;
  sort_order?: SortOrder;
}

export interface AdminUserListResponse {
  items: AdminUserSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  generated_at: string;
}

export interface AdminUserStatisticsResponse {
  total_users: number;
  active_users: number;
  locked_users: number;
  deleted_users: number;
  users_created_today: number;
  users_created_last_7_days: number;
  users_created_last_30_days: number;
  active_last_24_hours: number;
  active_last_7_days: number;
  generated_at: string;
}

export interface AdminUserMutationResponse {
  user: AdminUserDetail;
  audit_event: Record<string, unknown>;
}

export interface AdminUserCreatePayload {
  full_name: string;
  email: string;
  role: AdminRole;
  password?: string;
  temporary_password?: string;
  email_verified?: boolean;
  current_quota?: Record<string, unknown> | null;
}

export interface AdminUserUpdatePayload {
  full_name?: string;
  email?: string;
  email_verified?: boolean;
  permissions_override?: string[];
}

export interface AdminPasswordResetResponse {
  user_id: string;
  temporary_password: string;
  password_reset_required: boolean;
  updated_at: string;
  audit_event: Record<string, unknown>;
}
